### DISCUSSION TITLE
Turning an inky function into a variable (getting the players name, then turning it into a variable)

### DISCUSSION THREAD

**Discussion by Inkynoob739** [inkle/ink#773] [Writing in ink] [answered]
Hello, I am a total noob at inky so you might need to teach me some things.

I am trying to use inky to activate an external function that will ask the player's name for the story. I used this article to get the basics: [Article](https://wildwinter.medium.com/ink-tutorial-asking-the-players-name-5170d05c4ea8)
This worked just fine, until I added multiple calls, which made the dialog box appear multiple times. Then, I saw @joethephish say this:

"This is up to the individual game to implement. Ink's primary goal is to be a single layer within the technology stack of a game. The fact that it's also increasingly useable as a standalone IF language is sort of a side-effect, but we'd always expect you to customize the interface, whether it's by modifying the default Inky template, or by creating something in Unity. So, what you'd need to do is pop up a dialogue box to ask for the player's name, and then poke it in directly by setting an ink variable from the game side."
Originally posted by @joethephish in [#139 (comment) ](https://github.com/inkle/ink/issues/139#issuecomment-236834089)

How would "I poke it into a variable?"

_Note: I am not using Unity._

**Comment by Selsynn** [comment, marked as answer]
Hi, 

It would depend on the language you're using. All the time, there is a way to write directly to the variable itself. 
For example in js I will do `story.variablesState.$("namePCInk", namePCJs);` supposing the the first is the ink name of the variable `namePCInk` and I fill it with the value of `namePCJs` that is the name and the value of my js variable I already have filled with the current element I want to have.

All I would have to tell you, is either tell us what the port you are using (with what language?) and to look at the documentation.
You especially want to part where the port is explaining how to do this part of the official explanation https://github.com/inkle/ink/blob/master/Documentation/RunningYourInk.md#settinggetting-ink-variables

**Reply by Inkynoob739** [reply]
Hello.

Again, this is quite confusing for me. If the language is the code, I mostly try to stick to Ink but I can use js if I have to. 

I can sort of understand the code there but how would it ask for the players name? I want the player to open the html file and be able to type in their name, if possible. Then the game will use their name in the story. In the article I read, it worked until I called the variable multiple times. Is there a fix to that?

Thank you.

**Reply by Selsynn** [reply]
The answer are used to the web export (so HTML /  JS). For any other export, you'll have to do it by your own hand.

To create an input for the player to give the name, just pick any tutorial about doing a bit of HTML/JS 

1)First create an HTML input, and write it in your ink file, whenever the element should show up.
```html
<input id="namePC" type="text"></input>
```
It won't show a input box in inky, but it will show in the web export. 

But there it's not yet used , and the user can enter anything but the game can't pick it up. 

2) Getting the value from the input, by using a tag (The new tag will be "#PICK_NAME" for example) for example and handling it in the js part (read more about it here https://github.com/inkle/ink/blob/master/Documentation/RunningYourInk.md#marking-up-your-ink-content-with-tags )
At this part you have the beginning of the handling of the tag, and you just add another if element like the last one that is the CLEAR  and restart one https://github.com/inkle/inky/blob/43c89a629238eba3b3cfdfd5199680f29957c506/app/export-for-web-template/main.js#L129
```js
 // PICK_NAME
if( tag == "PICK_NAME" ) {
  var namePCJs = document.querySelector('#namePC').value;
  story.variablesState.$("namePCInk", namePCJs);
}
```
The `var namePCJs = document.querySelector('#namePC').value;` puts the value of the HTML input to the JS var. 
The `story.variablesState.$("namePCInk", namePCJs);` is where it goes from the JS to the ink. 

Warning; you have to call the tag AFTER the user has finished editing it. It means after the user validate it's choice for example. 
And it will be evaluated at the end of that bloc, so ink won't know the value until after the next choice.

Following that you can use it in ink. 
mini ink example
```ink 
VAR namePCInk=""
<input id="namePC" type="text"></input>
 * Validate  
 # PICK_NAME
 If you are sure, go for it !
 (don't change it now, it's too late !)

 ** Sure
 You have chosen the name "{namePCInk}"
```

It's not a great solution, but it will work for some input like this. I wouldn't recommend doing too much, and there are some rules to follow to avoid most errors, but if you stick with some really easy thing like that, it should work. Let me know if you need any more help. 

If you are stumbling and not understanding what you did wrong, try to have the smallest part that still acts wrongly so that we can look at your code and help you.

**Reply by Inkynoob739** [reply]
Okay, I think this makes sense. I will try it and let you know how it goes. Thank you so much!

**Reply by Inkynoob739** [reply]
Worked like a charm. Thank you for all your help!

**Reply by only-miscellany** [reply]
Hello!

I registered specifically so I could leave a comment here, so I am brand new to the site and hope I'm not making an etiquette faux pas by chiming in. I'm also brand new to ink and inky and trying to accomplish this same feat (allowing the player to choose their own name), which seems impossible to do smoothly without some other coding language(s) at your disposal, which is frustrating, because it seems like such a basic thing to me.

I've tried the external function method referenced earlier, and I had the same problem of needing to supply the input repeatedly, every time the function was called. It's workable, but it doesn't make for a great player experience.

I've tried the method outlined in this thread, following the instructions as closely as I can, and I haven't been able to get the input to populate properly. I'm not sure what I'm doing wrong, and I don't know enough about what I'm doing to be able to figure it out myself, or even where to start looking for the information I need to be able to do that. I've been scouring the internet for other alternatives to see if they might work better for me, but I just haven't found anything that both answers this question/solves this problem and is intelligible to me. You'd think that inky would have this function built in, since it's for creating branching narratives, and most players are going to want to name their hero. At least I would think it should just be part of the functionality of the thing itself. But I know enough to know that I don't know enough to be able to determine whether that actually makes sense or not. 

I want to learn, so I'm not asking to have anything done for me, but I'm not ready to master HTML/JS in order to understand how to perform the seemingly simple task of collecting input from a player and getting ink to return or print that input multiple times over the course of a story. Please help! Is there anything anyone would need from me in order to facilitate said help?

**Reply by Selsynn** [reply]
Hi @only-miscellany 

Don't worry, it's okay to ask questions. Else how can you learn anything? 

One thing to note, is that I always see ink as a tool, not as a game engine per se. 
It's created (I think) to go with another engine, either the inkjs or the unity plugin. In both case, getting an input from the user is something very easy to do (in html/js for the inkjs, or in c# with unity for the unity plugin) 


The inky editor is something different, it's another tool to quickly see what you are doing. It's imperfect, of course. The problem lays that it can't do the pure "engine" works, as it's engine-independent. You should be able to inject to ink created with inky to the unity plugin or to the inkjs website skeleton, and it should work without problem. 

The picking a name is something that is totally different in both language. Sure, we could do something in ink that would call that method, but that's 1) super complicated to have it generated correctly 2) almost sure everyone will override to make the transition, style, and whatnot they want. 

Coming back to the problem, 
1) are you in inky, in the editor itself when trying the solution? The inky editor, last time I checked, will not show the input (because, there is no way to tell you how to display some html element (the `<input>` itself is html ) ) so it will ignore the input. 
2) I'm now supposing that you are exporting to web from inky and using that to see it. In the project where you save it, you should have the main.js where you did the modification. Can you show it? It should be pretty standard, with just that modification, so easy to follow for me, as you don't seem to have modified anything else. Can you also try to have the smallest inky script where you repeat your problem, for example, setting the name, confirming it, and trying to display it, and having it blank so needing to display the form again.

**Reply by only-miscellany** [reply]
Thank you so much for getting back to me.

I for sure agree, ink is more of a tool in the context of designing full-fledged games. I'm keeping things simple right now and sticking with text-only content, so it actually has most of the functionality I need built into it. I've taken a look at learning to use unity with inky, and I'm hoping to be ready to dive into that soon, but it's not something I need for this project, cause I'm just trying to write a book where the reader can input the name of the main character.

To answer your questions:

1. When trying the solution(s), I've followed the steps necessary to upload the exported project to itch.io and just checked it out to see what's happening, cause you're right, inky can't collect or display input the way we're trying to do it here.
2. You suppose correctly. I just want to make sure I understand what you're asking for as far as showing it goes. Should I get some screenshots of the snippet of the main.js file where the modification is (Right again; I haven't modified anything else.), and then the finished project in operation so you can see what it's doing?

Thanks again for your attention here; I really appreciate the help.

**Reply by Selsynn** [reply]
@only-miscellany yes, get screenshot/copy paste of your main.js (if you copy the code, have it between the "code" tag, it will be easier to read.) and where you are using it in your ink file. (and the logic around it)

**Reply by only-miscellany** [reply]
Thanks for clarifying, and sorry it's taken me so forever to get back to you on this. Things have been a little crazy lately; I'm sure you know how it is.

Here are the relevant lines in ink:
![Inky File](https://user-images.githubusercontent.com/121378671/212501648-954bb124-8d96-4c0c-a12d-bcab7001dcfd.PNG)

And this is what the snippet of my main.js file looks like after I modified it:
![Inky Main JS File Modification](https://user-images.githubusercontent.com/121378671/212501649-93b0b889-0623-4844-a471-338cff3807a9.PNG)

And this is what it does when uploaded to itch.io and the player goes through the process of selecting a name:
![Inky Behavior in Browser](https://user-images.githubusercontent.com/121378671/212501647-88dddbdd-9433-4e2b-a66d-9cf6e07f419e.PNG)

I'm sure I've just stuck something in the wrong spot in the main.js file or don't know enough about syntax to make sure I have all the punctuation mark thingies in the right places, or something. It's probably something really simple, and I'm just blind to it because I honestly have no idea what I'm doing. So I really appreciate you being willing to take the time and give this your attention. Please let me know if there's anything else you need from me to be able to see the whole picture.

**Reply by Selsynn** [reply]
if you are okay with me coming back to you so late, I just spotted the error.

it's where the if is placed in the js. 
A little explication is required so you won't make the same mistake. 
the if structure I had you change is composed of the key word `if`, of a condition `(condition)` and if that condition is true, then, it apply what is inside the `{ action to do }`

for example in the idea 
```
var animal = "Cat";
if(animal== "Dog" ){
 Bark();
}
if (animal == "Cat"){
 TakeOverTheWorld()
}
```
That would works,( In the condition the double equal is a test for equality) except that Bark() and TakeOverTheWorld() are not defined, that just to give you the idea.
In natural language, I would translate as
```
There is a animal.
If the animal is a dog, then do the action Bark.
If the animal is a cat, then do the action TakeOverTheWorld.
```

What you did in you code is that however, still in that idea/notation (the double vertical means a OR)

```
var animal = "Cat";
if(animal== "Dog" || animal == "Rabbit" ){
  if(animal== "Dog" ){
   Bark();
  }
  if (animal == "Rabbit"){
   BeFluffy();
    if (animal == "Cat"){
     TakeOverTheWorld()
    }
  }
}
```

That would means in natural language

```
There is a animal.
If the animal is a dog or a rabbit, 
then 
if the animal is a dog then Bark  -- end if the animal is a dog
if the animal is a rabbit 
then 
BeFluffy  
if the animal is a cat then TakeOverTheWorld -- end if the animal is a cat
-- end if the animal is a rabbit
-- end if the if the animal is a dog or a rabbit
```

The problem if the animal is a cat, it does not enter the first condition, so it won't ever check the internal condition.
Just make sure that you are doing outside of any `tag == SPECIFICNAME` because you are testing for another specific value for tag. 

you can always write comments in js like this `//` I would suggest to write the end of the if in the closing bracket to keep track for the first time you are modifying it. 

And sorry for the late reply, I was super busy at work.

**Reply by only-miscellany** [reply]
Oh-kay! I think I just figured out what you're talking about with the double vertical and the placement in the js, because there isn't a double vertical in the code that I added, but there is one in the existing code in the js document, and I was really confused for a second. Or several minutes of reading and re-reading. So... I probably need to move the modification somewhere outside of some brackets which it's currently in so it stops being a part of the conditional logic that includes an or? And that's basically what you said, just in different words about checking the internal condition.

Thank you for breaking all that down so basically for me, that was really helpful! No worries on the reply time; I really appreciate you volunteering your spare time to teach me stuff. Thank you also for the tip about comments for tracking modification, because now I find that I don't remember which bracket goes where, but I'll play around with stuff and see what I can get to work so you don't have to try to physically point to me, "here's where you should put this code."

**Reply by only-miscellany** [reply]
Yes! I moved some stuff around, and after a few tries, I got things situated, and it's working! I just stuck the thing at the beginning of all the "ifs" instead of the end, and that seems to make a whole lot more sense to everybody. Thank you so much for your help!
